linera-chain/src/data_types.rs

```diff
@@ -587,16 +587,62 @@ impl<'a> SignatureAggregator<'a> {
     }
 }
 
+fn is_ordered(values: &[(ValidatorName, Signature)]) -> bool {
+    values.windows(2).all(|pair| pair[0].0 <= pair[1].0)
+}
+
+impl<'de> Deserialize<'de> for Certificate {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct CertificateVisitor;
+
+        impl<'de> serde::de::Visitor<'de> for CertificateVisitor {
+            type Value = Certificate;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                formatter.write_str("struct Certificate")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: serde::de::SeqAccess<'de>,
+            {
+                let value: HashedValue = seq
+                    .next_element()?
+                    .ok_or_else(|| serde::de::Error::invalid_length(0, &self))?;
+
+                let round: RoundNumber = seq
+                    .next_element()?
+                    .ok_or_else(|| serde::de::Error::invalid_length(1, &self))?;
+
+                let mut signatures: Vec<(ValidatorName, Signature)> = seq
+                    .next_element()?
+                    .ok_or_else(|| serde::de::Error::invalid_length(2, &self))?;
+
+                Ok(Certificate::new(value, round, &mut signatures))
+            }
+        }
+
+        deserializer.deserialize_seq(CertificateVisitor)
+    }
+}
+
 impl Certificate {
     pub fn new(
         value: HashedValue,
         round: RoundNumber,
-        signatures: Vec<(ValidatorName, Signature)>,
+        signatures: &mut Vec<(ValidatorName, Signature)>,
```

Comment:
Why not take ownership, instead of clone?

```suggestion
        mut signatures: Vec<(ValidatorName, Signature)>,
```

Reply:
Indeed `&mut` is not what we want.

Reply:
Ah, right. Because we don't want to modify the original vector, but rather transfer ownership. Makes sense! Thank you, I'll change it